expression/constant_fold.go

```diff
@@ -78,6 +79,51 @@ func ifNullFoldHandler(expr *ScalarFunction) (Expression, bool) {
 	return expr, isDeferredConst
 }
 
+func caseWhenHandler(expr *ScalarFunction) (Expression, bool) {
+	args, l := expr.GetArgs(), len(expr.GetArgs())
+	isDeferredConst, hasNonConstCondition := false, true
+	for i := 0; i < l-1; i += 2 {
+		foldedArg, isDeferred := foldConstant(args[i])
+		expr.GetArgs()[i] = foldedArg
+		isDeferredConst = isDeferredConst || isDeferred
+		if _, isConst := foldedArg.(*Constant); isConst && hasNonConstCondition {
+			// If the condition is const and true, and the previous conditions
+			// has no expr, then the folded execution body is returned, otherwise
+			// the arguments of the casewhen are folded and replaced.
+			val, isNull, err := args[i].EvalInt(expr.GetCtx(), chunk.Row{})
+			if err != nil {
+				return expr, false
+			}
+			if val != 0 && !isNull {
+				foldedExpr, isDeferred := foldConstant(args[i+1])
+				foldedExpr.GetType().Decimal = expr.GetType().Decimal
+				isDeferredConst = isDeferredConst || isDeferred
+				return foldedExpr, isDeferredConst
+			}
+		} else {
+			hasNonConstCondition = false
+		}
+		expr.GetArgs()[i+1], isDeferred = foldConstant(args[i+1])
+		isDeferredConst = isDeferredConst || isDeferred
+	}
+
+	if l%2 == 1 && hasNonConstCondition {
+		// If the number of arguments in casewhen is odd, and the previous conditions
+		// is const and false, then the folded else execution body is returned. otherwise
+		// the execution body of the else are folded and replaced.
+		foldedExpr, isDeferred := foldConstant(args[l-1])
+		foldedExpr.GetType().Decimal = expr.GetType().Decimal
```

Comment:
Hi, I rethink this line, use `BuildCastFunction` will be better.
We can remove this line and `return BuildCastFunction(ctx, foldedExpr, expr.GetType()) `.

Reply:
I've modified as your suggestion. PTAL @XuHuaiyu